torch/csrc/jit/codegen/cuda/lower_insert_syncs.cpp

```diff
@@ -0,0 +1,192 @@
+#include <torch/csrc/jit/codegen/cuda/ir_iostream.h>
+#include <torch/csrc/jit/codegen/cuda/lower_utils.h>
+
+#include <torch/csrc/jit/codegen/cuda/lower_insert_syncs.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+
+namespace {
+
+class LocalSyncInserter final : OptOutDispatch {
+ public:
+  static void InsertSyncs(Expr* expr) {
+    LocalSyncInserter opex(expr);
+  }
+
+  bool hasWARrace() const {
+    return war_sync_;
+  }
+
+ private:
+  void handle(Expr* expr) final {
+    if (ir_utils::isTVOp(expr)) {
+      // For this SyncInserter
+      (!initial_sync_) ? hasOutputSmemExpr(expr, initial_)
+                       : hasInputSmemExpr(expr, final_);
+
+      // For parent SyncInserter
+      hasOutputSmemExpr(expr, all_smem_outputs_);
+      hasInputSmemExpr(expr, all_smem_inputs_);
+    } else {
+      OptOutDispatch::handle(expr);
+    }
+  }
+
+  void handle(kir::IfThenElse* ite) final {
+    for (auto expr : ite->thenBody().exprs()) {
+      handle(expr);
+    }
+    for (auto expr : ite->elseBody().exprs()) {
+      handle(expr);
+    }
+  }
+
+  void handle(kir::ForLoop* fl) final {
+    for (auto expr : fl->body().exprs()) {
+      last_op_sync_ = false;
+      if (expr->getExprType().value() == ExprType::Sync) {
+        initial_sync_ = true;
+        final_.clear();
+      } else if (expr->getExprType().value() == ExprType::ForLoop) {
+        // Recursively handle nested for-loop
+        LocalSyncInserter child(expr);
+        const auto& child_inputs = child.all_smem_inputs_;
```

Comment:
the recursion is fine, but if LocalSyncInserter is used recursively we should only use its public interface during recursion (rater than peeking at private members)